server.py

```diff
@@ -1,90 +1,385 @@
+    """
+    start_time = time.time()
     mode = request.form['mode']
+    text = request.form['text'].strip()
+    selected = request.form['lang']
+    notes = ""
+
+    sentence = text
+
+    language = "eng"
+    if selected == "Spanish":
+        language = "spa"
+
+    elif selected == "Chinese":
+        language = "cmn"
+
+    # -------------------------- Language detector ---------------------------------
+
+    lang_detect = detect(text)
+    if lang_detect not in ['en', 'es', 'zh-cn', 'zh-tw']:
+        return jsonify({'text': "Language not found", 'entities': [], 'notes': "Language not supported."})
+
+    english = ((lang_detect == 'en') and (language == "eng"))
+    spanish = ((lang_detect == 'es') and (language == "spa"))
+    chinese = ((lang_detect[0:2] == "zh") and (language == "cmn"))
+
+    if not (english or spanish or chinese):
+        selected = "Chinese"
+        if lang_detect == "en":
+            selected = "English"
+            language = "eng"
+        elif lang_detect == "es":
+            selected = "Spanish"
+            language = "spa"
+
+        notes = "Your selected language does not seem to match the language of the text." \
+                " We will be assuming that the text is in " + selected + "."
+
+    # =====================================================================================
+    # Stanford CoreNLP
+    # =====================================================================================
+
+    nlp = StanfordCoreNLP('http://localhost:9000')
+
+    properties = {'annotators': 'tokenize,ssplit',
+                  'outputFormat': 'json'}
+
+    if language == 'cmn':
+        properties['customAnnotatorClass.tokenize'] = 'edu.stanford.nlp.pipeline.ChineseSegmenterAnnotator'
+        properties['tokenize.model'] = 'edu/stanford/nlp/models/segmenter/chinese/ctb.gz'
+        properties['tokenize.sighanCorporaDict'] = 'edu/stanford/nlp/models/segmenter/chinese'
+        properties['tokenize.serDictionary'] = 'edu/stanford/nlp/models/segmenter/chinese/dict-chris6.ser.gz'
+        properties['tokenize.sighanPostProcessing'] = 'true'
+        properties['ssplit.boundaryTokenRegex'] = urllib.quote_plus('[!?]+|[。]|[！？]+')
+    elif language == 'spa':
+        properties['tokenize.language'] = 'es'
 
-    text = request.form['text']
-    logger.info( 'text-received: %s' % text )
+    output = nlp.annotate(text, properties=properties)
 
-    text = text.encode('ascii', 'ignore').strip()
-    text = sent_tokenize( text )
-    text = [ word_tokenize(t.strip()) for t in text ]
-    logger.info( 'text after ssplit & tokenize: %s' % str(text) )
+    pprint.pprint(output)
 
+    text_array = []
+    text_to_offset = {}
+    sentences = output['sentences']
+    for sent in sentences:
+        new = []
+        tokens = sent['tokens']
+        for word in tokens:
+            new.append(word['originalText'])
```

Comment:
I guess I didn't explain it clearly. **It is to display the "originalText" but to send the "word" to our annotator (fofe-ner).** Please pay attention to what I circled in the screenshot. '(' is escaped as '-LRB-', which stands for Left Round Bracket.

Because "word" may not have the same length as "originalText", the offset computed from "word" is not right. Make advantage of the offsets returned by CoreNLP.

![image](https://user-images.githubusercontent.com/6295455/27758722-3b0aa444-5dea-11e7-8d7b-a2969e6ed224.png)

![image](https://user-images.githubusercontent.com/6295455/27758716-1f2b77f8-5dea-11e7-8588-c82cf28c04bb.png)